### ISSUE TITLE
net: vm_fault assert comming from route_cache::lookup()

### ISSUE THREAD

**Issue by vladzcloudius** [cloudius-systems/osv#480]
tree: master (000ce9a2cb638d00ecafe77780e21ad57331f239) with my networking statistics patches that have been recently integrated into the master.
Setup: ran (OSv) - sif, 1G default interface (10.0.0.0/24 subnet).
Build command:  `make -j5 modules=cassandra`
OSv command: `sudo scripts/run.py -m16G -c4 -nv -b bridge0 --mac a0:48:1c:d6:c4:c2`
Workload: the assert had happen when the system was idle after running cassandra tests. 
Note: While the system was idle there are still packets passing on the NIC since it's a default network (e.g. broadcasts).

```
The assert:
page fault outside application, addr: 0x0000000000000000
[registers]
RIP: 0x0000000000264cf0 <ip_output(mbuf*, mbuf*, route*, int, ip_moptions*, inpcb*)+3840>
RFL: 0x0000000000010246  CS:  0x0000000000000008  SS:  0x0000000000000010
RAX: 0xfffffffffffff8c8  RBX: 0xffffa0015e56cdc4  RCX: 0x0000000000000000  RDX: 0x0000000000000000
RSI: 0xffff800100331560  RDI: 0xffff80010948a010  RBP: 0xffff800103ea1cc0  R8:  0x0000000000e2c0a0
R9:  0x0000000000e2c0a0  R10: 0x0000000000000000  R11: 0x00000000618e7200  R12: 0xffff800103ea1b7c
R13: 0x0000000000e2c0a0  R14: 0x0000000000000000  R15: 0x00000000618e7200  RSP: 0xffff800103ea1b00
Aborted

[backtrace]
0x000000000032801f <???+3309599>
0x00000000003291b3 <mmu::vm_fault(unsigned long, exception_frame*)+147>
0x0000000000388e79 <page_fault+105>
0x0000000000387d66 <???+3702118>
0x0000000000261082 <???+2494594>
0x00000000002614e7 <icmp_error(mbuf*, int, int, unsigned int, int)+647>
0x000000000026c477 <udp_input(mbuf*, int)+1495>
0x0000000000263679 <ip_input+57>
0x00000000002507f1 <netisr_dispatch_src+417>
0x00000000002487ce <???+2394062>
0x00000000002507f1 <netisr_dispatch_src+417>
0x0000000000358e4a virtio::net::receiver()+506
0x000000000035913e <???+3510590>
0x00000000003e28ea <thread_main_c+26>
0x0000000000388ce5 <???+3706085>

Backtrace:
(gdb) bt
#0  0x00000000003f8512 in cli_hlt () at /home/vladz/vladz-github/osv/arch/x64/processor.hh:242
#1  halt_no_interrupts () at /home/vladz/vladz-github/osv/arch/x64/arch.hh:48
#2  osv::halt () at /home/vladz/vladz-github/osv/core/power.cc:34
#3  0x00000000002235f5 in abort (fmt=fmt@entry=0x6028cd "Aborted\n") at /home/vladz/vladz-github/osv/runtime.cc:150
#4  0x0000000000223620 in abort () at /home/vladz/vladz-github/osv/runtime.cc:117
#5  0x0000000000328020 in mmu::vm_sigsegv (addr=<optimized out>, ef=0xffff800103e812f8) at /home/vladz/vladz-github/osv/core/mmu.cc:1194
#6  0x00000000003291b4 in mmu::vm_fault (addr=<optimized out>, addr@entry=0, ef=ef@entry=0xffff800103e812f8) at /home/vladz/vladz-github/osv/core/mmu.cc:1216
#7  0x0000000000388e7a in page_fault (ef=0xffff800103e812f8) at /home/vladz/vladz-github/osv/arch/x64/mmu.cc:38
#8  <signal handler called>
#9  operator++ (this=<synthetic pointer>) at /home/vladz/vladz-github/osv/external/x64/gcc.bin/usr/include/c++/4.8.2/bits/stl_list.h:235
#10 __distancestd::_List_const_iterator<silly_rtable::silly_rtable_entry > (__last=..., __first=<error reading variable: Cannot access memory at address 0x10>) at /home/vladz/vladz-github/osv/external/x64/gcc.bin/usr/include/c++/4.8.2/bits/stl_iterator_base_funcs.h:82
#11 distancestd::_List_const_iterator<silly_rtable::silly_rtable_entry > (__last=..., __first=...) at /home/vladz/vladz-github/osv/external/x64/gcc.bin/usr/include/c++/4.8.2/bits/stl_iterator_base_funcs.h:118
#12 size (this=0xffffa0010948a4a0) at /home/vladz/vladz-github/osv/external/x64/gcc.bin/usr/include/c++/4.8.2/bits/stl_list.h:874
#13 add (r=..., n=16777215, a=1765997891, this=0xffffa0010948a4a0) at /home/vladz/vladz-github/osv/bsd/sys/net/routecache.hh:110
#14 lookup (ret=0xffff800103ea1bc0, fibnum=<optimized out>, dst=0xffff800103ea1b7c) at /home/vladz/vladz-github/osv/bsd/sys/net/routecache.hh:160
#15 ip_output (m=m@entry=0xffffa0015e56cd00, opt=opt@entry=0x0, ro=0xffff800103ea1b60, ro@entry=0x0, flags=flags@entry=0, imo=<optimized out>, imo@entry=0x0, inp=inp@entry=0x0) at /home/vladz/vladz-github/osv/bsd/sys/netinet/ip_output.cc:260
#16 0x0000000000261083 in icmp_send (opts=0x0, m=0xffffa0015e56cd00) at /home/vladz/vladz-github/osv/bsd/sys/netinet/ip_icmp.cc:911
#17 icmp_reflect (m=m@entry=0xffffa0015e56cd00) at /home/vladz/vladz-github/osv/bsd/sys/netinet/ip_icmp.cc:877
#18 0x00000000002614e8 in icmp_error (n=n@entry=0xffffa001107e8500, type=type@entry=3, code=<optimized out>, code@entry=3, dest=dest@entry=0, mtu=mtu@entry=0) at /home/vladz/vladz-github/osv/bsd/sys/netinet/ip_icmp.cc:342
#19 0x000000000026c478 in udp_input (m=0xffffa001107e8500, off=<optimized out>) at /home/vladz/vladz-github/osv/bsd/sys/netinet/udp_usrreq.cc:557
#20 0x000000000026367a in ip_input (m=<optimized out>) at /home/vladz/vladz-github/osv/bsd/sys/netinet/ip_input.cc:774
#21 0x00000000002507f2 in netisr_dispatch_src (proto=1, source=<optimized out>, m=0xffffa001107e8500) at /home/vladz/vladz-github/osv/bsd/sys/net/netisr.cc:769
#22 0x00000000002487cf in ether_input_internal (m=0xffffa001107e8500, ifp=0xffff900100d99000) at /home/vladz/vladz-github/osv/bsd/sys/net/if_ethersubr.cc:741
#23 ether_nh_input (m=0xffffa001107e8500) at /home/vladz/vladz-github/osv/bsd/sys/net/if_ethersubr.cc:754
#24 0x00000000002507f2 in netisr_dispatch_src (proto=9, source=<optimized out>, m=0xffffa001107e8500) at /home/vladz/vladz-github/osv/bsd/sys/net/netisr.cc:769
#25 0x0000000000358e4b in virtio::net::receiver (this=0xffff800103e7c040) at /home/vladz/vladz-github/osv/drivers/virtio-net.cc:495
#26 0x000000000035913f in operator() (__closure=<optimized out>) at /home/vladz/vladz-github/osv/drivers/virtio-net.cc:234
#27 std::_Function_handler<void(), virtio::net::net(pci::device&)::__lambda12>::_M_invoke(const std::_Any_data &) (__functor=...) at /home/vladz/vladz-github/osv/external/x64/gcc.bin/usr/include/c++/4.8.2/functional:2071
#28 0x00000000003e28eb in main (this=0xffff800103e7c2d0) at /home/vladz/vladz-github/osv/core/sched.cc:940
#29 sched::thread_main_c (t=0xffff800103e7c2d0) at /home/vladz/vladz-github/osv/arch/x64/arch-switch.hh:137
#30 0x0000000000388ce6 in thread_main () at /home/vladz/vladz-github/osv/arch/x64/entry.S:113

(gdb)
```

**Comment by wkozaczuk** [comment]
The code changes that are part of the commit c9129eefecd9c78f86efe37a8ef2623a3ae6d6cf and the related conversation on the mailing list (see https://groups.google.com/g/osv-dev/c/I-AMAXmt6vI/m/HpJh5RyRl04J) involving @nyh suggest this issue indeed may be fixed.

The commit changes line 111 but the crash reported by this issue happens in 110 (one above). Would the crash in 110 be caused by the bug in line 111?

**Comment by wkozaczuk** [comment]
I have created this minimal example to verify that c9129eefecd9c78f86efe37a8ef2623a3ae6d6cf indeed fixed this issue:
```c++
#include <cstdio>
#include <list>

class silly_rtable {
    // Limit the number of routing entries cached to some reasonable small
    // limit. This is an O(N) data structure, after all...
    static constexpr unsigned max_entries = 10;
    struct silly_rtable_entry {
        unsigned int address;
        unsigned int netmask;
        silly_rtable_entry(unsigned int a, unsigned int n) :
            address(a), netmask(n) { }
    };
    std::list<silly_rtable_entry> entries;
public:
    void add(unsigned int a, unsigned int n) {
        while (entries.size() >= max_entries) {
            //entries.pop_back();
            entries.erase(entries.end());
        }
        entries.emplace_front(a, n);
    }
};

int main()
{
   silly_rtable srt;
   for (int i = 0; i < 15; i++) {
      printf("Adding the %dth entry\n", i);
      srt.add(1,2);
   }
}
```

and indeed it crashes before the fix and does not after. The stack trace looks different from what was reported 10 years ago:
```
OSv v0.57.0-166-g1836c75a
eth0: 192.168.122.15
Booted up in 186.37 ms
Cmdline: /route-cache
Adding the 0th entry
Adding the 1th entry
Adding the 2th entry
Adding the 3th entry
Adding the 4th entry
Adding the 5th entry
Adding the 6th entry
Adding the 7th entry
Adding the 8th entry
Adding the 9th entry
Adding the 10th entry
Adding the 11th entry
Adding the 12th entry
Adding the 13th entry
Adding the 14th entry
Assertion failed: !safemode_or_autounlink || node_algorithms::unique(to_insert) (/usr/include/boost/intrusive/bstree.hpp: insert_equal: 1065)

(gdb) bt
#0  0x000000004031ea32 in processor::cli_hlt () at arch/x64/processor.hh:247
#1  arch::halt_no_interrupts () at arch/x64/arch.hh:61
#2  osv::halt () at arch/x64/power.cc:29
#3  0x0000000040244e60 in abort (fmt=fmt@entry=0x405ecba0 "Assertion failed: %s (%s: %s: %d)\n") at runtime.cc:145
#4  0x0000000040244e89 in __assert_fail (expr=<optimized out>, file=<optimized out>, line=<optimized out>, func=<optimized out>) at runtime.cc:153
#5  0x000000004036db4d in boost::intrusive::bstree_impl<boost::intrusive::mhtraits<memory::page_range, boost::intrusive::set_member_hook<>, &memory::page_range::set_hook>, void, void, unsigned long, false, (boost::intrusive::algo_types)5, void>::insert_equal(memory::page_range&) (value=..., this=<optimized out>)
    at /usr/include/boost/intrusive/bstree.hpp:1065
#6  0x000000004037369b in boost::intrusive::bstree_impl<boost::intrusive::mhtraits<memory::page_range, boost::intrusive::set_member_hook<>, &memory::page_range::set_hook>, void, void, unsigned long, false, (boost::intrusive::algo_types)5, void>::insert_equal(memory::page_range&) (value=..., this=<optimized out>)
    at /usr/include/boost/intrusive/detail/rbtree_node.hpp:102
#7  boost::intrusive::multiset_impl<boost::intrusive::mhtraits<memory::page_range, boost::intrusive::set_member_hook<>, &memory::page_range::set_hook>, void, void, unsigned long, false, void>::insert(memory::page_range&) (value=..., this=<optimized out>) at /usr/include/boost/intrusive/set.hpp:748
#8  memory::page_range_allocator::insert<true> (this=<optimized out>, pr=...) at core/mempool.cc:625
#9  0x000000004036f905 in memory::page_range_allocator::free (this=this@entry=0x4089b6e0 <memory::free_page_ranges>, pr=<optimized out>, pr@entry=0x400000d87000)
    at core/mempool.cc:823
#10 0x00000000403704ab in memory::free_page_range_locked (range=0x400000d87000) at core/mempool.cc:1138
#11 memory::free_page_range (range=0x400000d87000) at core/mempool.cc:1146
#12 memory::free_large (obj=0x400000d87000) at core/mempool.cc:1159
#13 free (object=<optimized out>) at core/mempool.cc:1946
#14 free (object=<optimized out>) at core/mempool.cc:1928
#15 0x0000000000401b1f in std::__new_allocator<std::_List_node<silly_rtable::silly_rtable_entry> >::deallocate(std::_List_node<silly_rtable::silly_rtable_entry>*, unsigned long) ()
#16 0x00000000004019d6 in std::allocator_traits<std::allocator<std::_List_node<silly_rtable::silly_rtable_entry> > >::deallocate(std::allocator<std::_List_node<silly_rtable::silly_rtable_entry> >&, std::_List_node<silly_rtable::silly_rtable_entry>*, unsigned long) ()
#17 0x0000000000401858 in std::__cxx11::_List_base<silly_rtable::silly_rtable_entry, std::allocator<silly_rtable::silly_rtable_entry> >::_M_put_node(std::_List_node<silly_rtable::silly_rtable_entry>*) ()
#18 0x0000000000401797 in std::__cxx11::_List_base<silly_rtable::silly_rtable_entry, std::allocator<silly_rtable::silly_rtable_entry> >::_M_clear() ()
#19 0x0000000000401548 in std::__cxx11::_List_base<silly_rtable::silly_rtable_entry, std::allocator<silly_rtable::silly_rtable_entry> >::~_List_base() ()
#20 0x000000000040139c in std::__cxx11::list<silly_rtable::silly_rtable_entry, std::allocator<silly_rtable::silly_rtable_entry> >::~list() ()
#21 0x00000000004013d4 in silly_rtable::~silly_rtable() ()
#22 0x0000000000401204 in main ()
#23 0x00000000403babdd in osv::application::run_main (this=0x600000c39e10) at core/app.cc:462
#24 0x00000000403bad1f in __libc_start_main (main=0x4011a6 <main>) at core/app.cc:40
#25 0x00000000004010e5 in _start ()
```

On Linux it fails differently:
```
Adding the 0th entry
Adding the 1th entry
Adding the 2th entry
Adding the 3th entry
Adding the 4th entry
Adding the 5th entry
Adding the 6th entry
Adding the 7th entry
Adding the 8th entry
Adding the 9th entry
Adding the 10th entry
munmap_chunk(): invalid pointer
Aborted (core dumped)
```

We cannot be 100%, this is indeed the same issue that was fixed with c9129eefecd9c78f86efe37a8ef2623a3ae6d6cf, but given it is 10 years old and based on the description it was never easy to reproduce it, let us close it.
